imagecraft/models/volume.py

```diff
@@ -0,0 +1,104 @@
+# This file is part of imagecraft.
+#
+# Copyright 2025 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify it
+# under the terms of the GNU General Public License version 3, as published
+# by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
+# SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Volume configuration pydantic model."""
+
+import enum
+import re
+import typing
+import uuid
+from typing import Literal, cast
+
+from craft_application.models import CraftBaseModel
+from craft_application.models.constraints import UniqueList
+from craft_parts.utils.partition_utils import VALID_PARTITION_REGEX
+from pydantic import ByteSize, Field, StringConstraints
+
+from imagecraft.platforms import GPT_NAME_MAX_LENGTH, FileSystem, GptType
+
+VALID_VOLUME_NAME_REGEX = VALID_PARTITION_REGEX
+VALID_STRUCTURE_NAME_REGEX = VALID_PARTITION_REGEX
+
+StructureName = typing.Annotated[
+    str,
+    StringConstraints(
+        max_length=GPT_NAME_MAX_LENGTH, pattern=VALID_STRUCTURE_NAME_REGEX
+    ),
+]
+
+VolumeName = typing.Annotated[
+    str,
+    StringConstraints(pattern=VALID_VOLUME_NAME_REGEX, strict=True),
+]
+
+VOLUME_INVALID_MSG = (
+    "Volume names must only contain lowercase letters, numbers, "
+    "and hyphens, and may not begin or end with a hyphen."
+)
+
+
+class Role(str, enum.Enum):
+    """Role describes the role of given structure."""
+
+    SYSTEM_DATA = "system-data"
+    SYSTEM_BOOT = "system-boot"
+
+
+class StructureItem(CraftBaseModel):
+    """Structure item of the image."""
+
+    name: StructureName
+    id: uuid.UUID | None = None
+    role: Role
+    structure_type: GptType = Field(alias="type")
+    size: ByteSize = ByteSize(0)
+    filesystem: FileSystem
+    filesystem_label: str = ""
+
+    def __hash__(self) -> int:
+        return hash(self.name)
+
+    def __eq__(self, other: object) -> bool:
+        if type(other) is type(self):
+            return self.name == cast(StructureItem, other).name
+
+        return False
+
+
+class PartitionSchema(str, enum.Enum):
+    """Supported partition schemas."""
+
+    GPT = "gpt"
+
+
+class Volume(CraftBaseModel):
+    """Volume defining properties of the image."""
+
+    volume_schema: typing.Annotated[
+        Literal[PartitionSchema.GPT],
+        Field(default=PartitionSchema.GPT),
+    ] = Field(alias="schema")
+    structure: UniqueList[StructureItem] = Field(min_length=1)
```

Comment:
I am tempted to rename it "structures" (plural) and only expose the singular name to the user. I would make it clear in the code it holds a list of "structure items".